src/utils/merge-nodes-with-matching-flavors.ts

```diff
@@ -0,0 +1,166 @@
+import type {
+    EmptyStatement,
+    ImportDeclaration,
+    ImportDefaultSpecifier,
+    ImportNamespaceSpecifier,
+    ImportSpecifier,
+} from '@babel/types';
+
+import {
+    importFlavorRegular,
+    importFlavorType,
+    mergeableImportFlavors,
+} from '../constants';
+import type { MergeNodesWithMatchingImportFlavors } from '../types';
+import { getImportFlavorOfNode } from './get-import-flavor-of-node';
+
+type MergeableFlavor = typeof mergeableImportFlavors[number];
+function isMergeableFlavor(flavor: string): flavor is MergeableFlavor {
+    return mergeableImportFlavors.includes(flavor as MergeableFlavor);
+}
+
+function selectMergeableNodesByImportFlavor(
+    nodes: ImportDeclaration[],
+): Record<MergeableFlavor, ImportDeclaration[]> {
+    return nodes.reduce(
+        (groups, node) => {
+            const flavor = getImportFlavorOfNode(node);
+            if (isMergeableFlavor(flavor)) {
+                groups[flavor].push(node);
+            }
+            return groups;
+        },
+        {
+            [importFlavorRegular]: [] as ImportDeclaration[],
+            [importFlavorType]: [] as ImportDeclaration[],
```

Comment:
Typecasting works and is probably fine here, but normally I prefer to use a generic on the `reduce` call, like 

```ts
return nodes.reduce<{[key: typeof importFlavorRegular | typeof importFlavorType]: ImportDeclaration[]}>(
```

That's a bit clunky in this case, but it lets us avoid using `as`.  Go ahead and leave this as-is, though, just wanted to mention it.

Reply:
Did it, but slightly differently!